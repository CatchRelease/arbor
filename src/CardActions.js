import styled, { css } from 'react-emotion';

import Grid from './Grid';

const gridStyles = ({ children }) => css`
  grid-template-columns: repeat(${children.length}, 1fr);
`;

const CardActions = styled(Grid.withComponent('div'))`
  ${gridStyles};
`;

CardActions.defaultProps = {
  gridGap: 'small'
};

export default CardActions;
