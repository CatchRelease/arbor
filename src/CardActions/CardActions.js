import { css } from '@emotion/core';
import styled from '@emotion/styled';

import createWithComponent from '../utils/createWithComponent';
import Grid from '../Grid';

const gridStyles = ({ children }) => css`
  grid-template-columns: repeat(${children.length}, 1fr);
`;

const CardActions = styled(
  createWithComponent(Grid, 'div', {
    gridGap: 'smaller'
  })
)`
  ${gridStyles};
`;

export default CardActions;
