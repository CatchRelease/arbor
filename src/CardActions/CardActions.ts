import { ReactElement } from 'react';

import { css } from '@emotion/react';
import styled from '@emotion/styled';

import Grid from '../Grid';

type Children = ReactElement | ReactElement[];

const length = (children: Children) =>
  'length' in children ? children.length : 1;

const gridStyles = ({ children }: { children: Children }) => css`
  grid-template-columns: repeat(${length(children)}, 1fr);
`;

const CardActions = styled(Grid)`
  ${gridStyles};
`;

CardActions.defaultProps = {
  gridGap: 'smaller'
};

export default CardActions;
