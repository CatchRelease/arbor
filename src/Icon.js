import React from 'react';
import styled, { css, cx } from 'react-emotion';
import { color, display, fontSize, textAlign, space } from 'styled-system';
import { withTheme } from 'emotion-theming';
import PropTypes from 'prop-types';

const rotationStyles = ({ rotation }) =>
  rotation &&
  css`
    transform: rotate(${rotation}deg);
  `;

const StyledIcon = styled.i`
  ${color};
  ${display};
  ${fontSize};
  ${space};
  ${textAlign};
  ${rotationStyles};
`;

StyledIcon.propTypes = {
  ...color.propTypes,
  ...fontSize.propTypes,
  ...space.propTypes,
  ...textAlign.propTypes
};

const Icon = ({ name, theme, ...props }) => {
  const { iconFontPrefix } = theme;

  return (
    <StyledIcon
      className={cx(iconFontPrefix, `${iconFontPrefix}-${name}`)}
      {...props}
    />
  );
};

Icon.propTypes = {
  name: PropTypes.string.isRequired,
  rotation: PropTypes.oneOf(['90', '180', '270', null])
};

Icon.defaultProps = {
  rotation: null
};

export default withTheme(Icon);
