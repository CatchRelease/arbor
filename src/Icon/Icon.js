/** @jsx jsx */
import { ClassNames, css, jsx } from '@emotion/core';
import styled from '@emotion/styled';
import {
  color,
  display,
  fontSize,
  textAlign,
  space,
  textShadow,
  opacity
} from 'styled-system';
import { withTheme } from 'emotion-theming';
import PropTypes from 'prop-types';

const rotationStyles = ({ rotation }) =>
  rotation &&
  css`
    transform: rotate(${rotation}deg);
  `;

const StyledIcon = styled.i`
  ${color};
  ${display};
  ${fontSize};
  ${space};
  ${textAlign};
  ${rotationStyles};
  ${textShadow};
  ${opacity};
`;

StyledIcon.propTypes = {
  ...color.propTypes,
  ...fontSize.propTypes,
  ...space.propTypes,
  ...textAlign.propTypes,
  ...textShadow.propTypes,
  ...opacity.propTypes
};

const Icon = ({ name, theme, ...props }) => {
  const { iconFontPrefix } = theme;

  return (
    <ClassNames>
      {({ cx }) => (
        <StyledIcon
          className={cx(iconFontPrefix, `${iconFontPrefix}-${name}`)}
          {...props}
        />
      )}
    </ClassNames>
  );
};

Icon.propTypes = {
  /**
   * Name of the icon to render. Must be a supported icon loaded via an icon
   * font on the page. The font prefix will be prepended and can be customized
   * via the theme iconFontPrefix variable.
   */
  name: PropTypes.string.isRequired,

  /**
   * Icon font rotation. Rotation will be passed in to the CSS transform
   * property using rotate.
   */
  rotation: PropTypes.oneOf(['90', '180', '270', null]),

  /**
   * Theme used for styling the Icon.
   */
  theme: PropTypes.object.isRequired // eslint-disable-line react/forbid-prop-types
};

Icon.defaultProps = {
  rotation: null
};

export default withTheme(Icon);
