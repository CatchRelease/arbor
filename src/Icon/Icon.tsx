import { ComponentProps, FC, forwardRef, LegacyRef } from 'react';
import { ClassNames, css, Theme, withTheme } from '@emotion/react';
import styled from '@emotion/styled';
import {
  color,
  ColorProps,
  display,
  DisplayProps,
  fontSize,
  FontSizeProps,
  opacity,
  OpacityProps,
  space,
  SpaceProps,
  textAlign,
  TextAlignProps,
  textShadow,
  TextShadowProps
} from 'styled-system';

type Rotation = '90' | '180' | '270' | null;

const rotationStyles = ({ rotation }: { rotation?: Rotation }) =>
  rotation &&
  css`
    transform: rotate(${rotation}deg);
  `;

type RotationProps = {
  rotation?: Rotation;
};

export const StyledIcon = styled.i<
  ColorProps &
    DisplayProps &
    FontSizeProps &
    SpaceProps &
    TextAlignProps &
    RotationProps &
    TextShadowProps &
    OpacityProps
>`
  ${color};
  ${display};
  ${fontSize};
  ${space};
  ${textAlign};
  ${rotationStyles};
  ${textShadow};
  ${opacity};
`;

type InternalProps = ComponentProps<typeof StyledIcon> & {
  className?: string;
  iconRef?: LegacyRef<HTMLElement>;
  name: string;
  rotation?: Rotation;
  theme: Theme;
};

const Icon: FC<InternalProps> = ({
  className,
  iconRef,
  name,
  rotation = null,
  theme,
  ...props
}) => {
  const { iconFontPrefix } = theme;

  return (
    <ClassNames>
      {({ cx }) => (
        <StyledIcon
          className={cx(iconFontPrefix, `${iconFontPrefix}-${name}`, className)}
          ref={iconRef}
          rotation={rotation}
          {...props}
        />
      )}
    </ClassNames>
  );
};

const WithThemeIcon = withTheme(Icon);

const ForwardRefWithThemeIcon = forwardRef<
  HTMLElement,
  ComponentProps<typeof WithThemeIcon>
>((props, ref) => {
  return <WithThemeIcon {...props} iconRef={ref} />;
});

export type Props = ComponentProps<typeof ForwardRefWithThemeIcon>;

export default ForwardRefWithThemeIcon;
