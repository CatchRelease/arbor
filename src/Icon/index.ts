export { default, Props as IconProps } from './Icon';
