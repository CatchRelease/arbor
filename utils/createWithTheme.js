import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
