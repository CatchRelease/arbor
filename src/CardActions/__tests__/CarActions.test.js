import React from 'react';
import createWithTheme from '../../../utils/createWithTheme';

import CardActions from '../CardActions';
import Icon from '../../Icon';

describe('<CardActions', () => {
  it('renders the CardActions component properly', () => {
    const tree = createWithTheme(
      <CardActions>
        <Icon name="download" />
        <Icon name="chat-bubble-outline" />
      </CardActions>
    );

    expect(tree).toMatchSnapshot();
  });
});
