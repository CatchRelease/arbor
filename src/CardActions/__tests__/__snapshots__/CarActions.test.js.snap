// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardActions renders the CardActions component properly 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  grid-gap: 0.5rem;
  grid-template-columns: repeat(2, 1fr);
}

<div
  className="emotion-0 emotion-1"
>
  <i
    className="ar ar-download emotion-2"
  />
  <i
    className="ar ar-chat-bubble-outline emotion-2"
  />
</div>
`;
